File: 2nd_task/2nd task.cs
using System;

class Program
{
    static void Main()
    {
        Console.OutputEncoding = System.Text.Encoding.UTF8; // Установка кодировки для отображения символов в консоли

        Console.Write("Введите количество банок с водой: ");
        int bottles = int.Parse(Console.ReadLine());

        Console.Write("Введите количество пустых банок для обмена на 1 полную: ");
        int tradeNeeded = int.Parse(Console.ReadLine());

        int waterConsumed = bottles; // Количество выпитой воды
        int iterations = 0; // Количество итераций
        int emptyBottles = bottles; // Количество пустых банок воды

        // Процесс обмена пустых банок на полные
        if (bottles <= 0 || tradeNeeded <= 0)
        {
            Console.WriteLine("Ошибка! Неверно введенные данные!");
        }
        else
        {
            while (emptyBottles >= tradeNeeded)
            {
                int exchangedBottles = emptyBottles / tradeNeeded; // Количество полных банок, которые можно получить
                waterConsumed += exchangedBottles; // Добавляем выпитую воду
                emptyBottles = emptyBottles % tradeNeeded + exchangedBottles; // Обновляем количество пустых банок
                iterations++; // Увеличиваем счетчик итераций
                Console.WriteLine($"Итерация №{iterations}: Выпито {waterConsumed} банок воды. Пустых банок осталось: {emptyBottles}");
            }
        }

        // Вывод результатов
        Console.WriteLine($"Максимальное количество банок, которое можно выпить: {waterConsumed}");
        Console.WriteLine($"Общее количество итераций: {iterations}");
    }
}

File: laba2/2nd_task/2nd task.hs
import System.IO

main :: IO ()
main = do
    putStrLn "Введите количество банок с водой: "
    bottles <- readLn :: IO Int
    putStrLn "Введите количество пустых банок для обмена на 1 полную:"
    tradeNeeded <- readLn :: IO Int

    let waterConsumed = bottles -- Количество выпитой воды
        iterations = 0 -- Количество итераций
        emptyBottles = bottles -- Количество пустых банок воды

    -- Процесс обмена пустых банок на полные
    if bottles <= 0 then
        putStrLn "Ошибка! Неверно введенные данные!"
    else do
        let loop accEmptyBottles accWaterConsumed accIterations
             | accEmptyBottles < tradeNeeded = do
                 putStrLn $ "Максимальное количество банок, которое можно выпить: " ++ show accWaterConsumed
                 putStrLn $ "Общее количество итераций: " ++ show accIterations
             | otherwise = do
                 let exchangedBottles = accEmptyBottles `div` tradeNeeded -- Количество полных банок, которые можно получить
                     newWaterConsumed = accWaterConsumed + exchangedBottles -- Добавляем выпитую воду
                     newEmptyBottles = accEmptyBottles `mod` tradeNeeded + exchangedBottles -- Обновляем количество пустых банок
                     newIterations = accIterations + 1 -- Увеличиваем счетчик итераций
                 putStrLn $ "Итерация №" ++ show newIterations ++ ": Выпито " ++ show newWaterConsumed ++ " банок воды. Пустых банок осталось: " ++ show newEmptyBottles
                 loop newEmptyBottles newWaterConsumed newIterations
        loop emptyBottles waterConsumed iterations
